pkg/apis/onepassword/v1/onepasswordsecret_types.go

```diff
@@ -11,11 +11,34 @@ type OnePasswordItemSpec struct {
 	ItemPath string `json:"itemPath,omitempty"`
 }
 
+type OnePasswordItemConditionType string
+
+const (
+	// OnePasswordItemReady means the Kubernetes secret is ready for use.
+	OnePasswordItemReady OnePasswordItemConditionType = "Ready"
+)
+
+type OnePasswordItemCondition struct {
+	// Type of job condition, Completed.
+	Type OnePasswordItemConditionType `json:"type"`
+	// Status of the condition, one of True, False, Unknown.
+	Status metav1.ConditionStatus `json:"status"`
+	// Last time the condition transit from one status to another.
+	// +optional
+	LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty"`
+	// Human-readable message indicating details about last transition.
+	// +optional
+	Message string `json:"message,omitempty"`
+}
+
 // OnePasswordItemStatus defines the observed state of OnePasswordItem
 type OnePasswordItemStatus struct {
-	// INSERT ADDITIONAL STATUS FIELD - define observed state of cluster
 	// Important: Run "operator-sdk generate k8s" to regenerate code after modifying this file
 	// Add custom validation using kubebuilder tags: https://book-v1.book.kubebuilder.io/beyond_basics/generating_crd.html
+	Conditions []OnePasswordItemCondition `json:"conditions"`
+
+	// True when the Kubernetes secret is ready for use.
+	Ready *bool `json:"ready,omitempty"`
 }
```

Comment:
Just wondering...
Is it possible and would it help to have a pending status for a secret? I'm thinking about making a secret based on a large item. 

This suggestion may not make sense for a Kubernetes secret, therefore you can treat this comment as a non-blocker.

Reply:
Hi, apologize for jumping in to comment here rather in the forum, but I just happened to want to set up 1P in k8s literally today, and locking in an API might be forever... I highly recommend not using a `bool Ready` field, and instead use an enum for status. Using `string Status` would be more flexible in the future in case the workflow becomes more complicated in the future. I can't find a publicly available style guide for the reasoning, but [Google Cloud API's style guide](https://cloud.google.com/apis/design/design_patterns#bool_vs_enum_vs_string) is fairly close. Every project I worked on prohibited the use of boolean fields in APIs.

The most-canonical reference that should be looked at regarding status should be at [github.com/kubernetes/community](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status) for general guidance. 

A 1Password Secret is not unlike a Certificate request for a SSL cert.

One example that can be followed is Google's `ManagedCertificate`: 
https://github.com/GoogleCloudPlatform/gke-managed-certs/blob/b4bbbe4cf108ab28c224b340493a58d893bdf3c6/deploy/managedcertificates-crd.yaml#L122-L128

It can be seen that the type is `string` and the [troubleshooting guide](https://cloud.google.com/load-balancing/docs/ssl-certificates/troubleshooting#certificate-managed-status) lists these states like: `PROVISIONING` and `ACTIVE`. Having a state like `PROVISIONING` is important because it is a debugging aid to let a user know that the operator sees it and is able to modify the CRD at the very least. 

If there is a recommendation for something to start with, it would be a `string` with those two values. Perhaps also something like a `PROVISIONING_FAILED_PERMANENTLY` if there is some error like no permission to create or update the secret in the required namespace. Permission errors are likely to happen especially if the user wishes to trim back the RBAC permissions to only allow the operator to manage Secrets.

Reply:
No need to apologize, your input is actually very welcome! 🙌 

I agree with the inflexibility of using a bool. I pretty much slipped it in last-minute because I thought it was needed for [Terraform's `wait_for` construct](https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs/resources/manifest#using-wait_for-to-block-create-and-update-calls). But apparently, it is also possible to wait for specific conditions.

That pretty much removes the need for the `ready` or `phase` fields and allows us to stick to the convention documented in the [k8s guideline](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status):

> Some resources in the v1 API contain fields called phase, and associated message, reason, and other status fields. The pattern of using phase is deprecated. Newer API types should use conditions instead.

In other words, what do you say about changing it to this?
```go
type OnePasswordItemStatus struct {
	// Important: Run "operator-sdk generate k8s" to regenerate code after modifying this file
	// Add custom validation using kubebuilder tags: https://book-v1.book.kubebuilder.io/beyond_basics/generating_crd.html
	Conditions []OnePasswordItemCondition `json:"conditions"`
}
```

Reply:
To be honest, I'm not sure because it didn't seem possible to wait for specific conditions in the Terraform `wait_for`. Its docs say it is a map of field paths to regular expressions (`map(string)`). Trying to search for examples is difficult since the usage of the words condition and status is fairly wide. The problem I see is you need to select an item from an array which has the property `type: Ready` and then see if `status: True`. Seems like that is more of a path query than what Terraform path expressions support. Can you find an example? Even the [docs](https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs/resources/manifest#using-wait_for-to-block-create-and-update-calls) show an example that relies upon the assumption that index 0 of an array is what is needed.

I'd guess it would go with what the recommended convention and show an example workaround similar to [this for deployment failures](https://github.com/hashicorp/terraform-provider-kubernetes/issues/1455#issuecomment-1136522277) which doesn't seem complete since it missing the `--context` flag if users have multiple Kubernetes providers/contexts in use. I asked a question about conditions here: https://github.com/hashicorp/terraform-provider-kubernetes/issues/1750. On the other hand, without waiting for it, is the worst that can happen is a deployment keeps on trying to create the pod when the secret is not there yet?

Reply:
I don't think there currently is another way than using the `status.conditions[0]`. That's definitely not pretty and not extremely stable, but I feel that is more of a problem with the design of the k8s TF provider. After all, using `wait_for` will never guarantee that the resource ends up in the state defined in `wait_for`. It can only guarantee that it was at that state at some point after applying the manifest. That could very well be the state was still there from the last apply. Your issue on the TF provider might clear some stuff up here.

My proposal is to merge the conditions part of this MR now. Then at least that can be used. If at a later stage we find out that there is indeed a strong need to add a deprecated construct to the `status` field, we can always reconsider adding that.

Reply:
Seeing that the maintainers of `terraform-provider-kubernetes` responded that the next version of the provider will contain an option to wait for a specific condition, that completely removes the need for a `phase` or `ready` field. Hence, I am going forward with the following structure for `status`:
```go
type OnePasswordItemStatus struct {
	// Important: Run "operator-sdk generate k8s" to regenerate code after modifying this file
	// Add custom validation using kubebuilder tags: https://book-v1.book.kubebuilder.io/beyond_basics/generating_crd.html
	Conditions []OnePasswordItemCondition `json:"conditions"`
}
```

Reply:
Thanks again for chipping in! 🙌